src/base/emu-i386/simx86/protmode.c

```diff
@@ -392,27 +393,32 @@ int emu_ldt_write(unsigned char *paddr, uint32_t op, int len)
 
 void emu_check_read_pagefault(dosaddr_t addr)
 {
-	if (addr >= LOWMEM_SIZE + HMASIZE && !dpmi_read_access(addr)) {
+	if (emu_jmp_env) {
 		/* trigger an exception in DPMI */
 		TheCPU.err = EXCP0E_PAGE;
 		/* uncommitted page is never "present" */
 		TheCPU.scp_err = 4;
 		TheCPU.cr2 = addr;
-		longjmp(jmp_env, 0);
+		longjmp(*emu_jmp_env, 0);
 	}
+	dosemu_error("Invalid read from addr %#x, ptr %p\n",
+		     addr, MEM_BASE32(addr));
+	leavedos_main(1);
 }
 
 /* callers must make sure that addr and addr+len-1 are on the same page */
-int emu_check_write_pagefault(dosaddr_t addr, uint32_t op, int len)
+void emu_check_write_pagefault(dosaddr_t addr, uint32_t op, int len)
 {
-	if (addr >= LOWMEM_SIZE + HMASIZE && !dpmi_write_access(addr)) {
+	if (emu_jmp_env) {
```

Comment:
Now you made jmpbuf a global flag
with malloc... But what if someone else
calls here when you didn't do
```
free(emu_jmp_env);
emu_jmp_env = NULL;
```
? Then you will still jump to
no-man's land, so you always
need to make sure the flag is
freed and cleared...
Why just exposing some API
to public, if all it does when
called from the outside, is
leavedos?
You could also pass a call-back
to these functions, that will be
called in the case of a fault.
Do longjmp from the callback.
But, if there are no other users,
I just wonder what's the deal?

Reply:
I've redone it using a callback, really signal()-style interface since it's like registering a simulated SIGSEGV handler.